File: pkg/wbjp/history/history.go
// Package history はスイング売買のスクリーニング履歴（wbjp screen の結果）を積む。
//
// 置き場は state/wbjp/history/screen/（wbcore/history.Store）。1 回の screen が 1 ファイル。
// 合成後の意見がある銘柄は**閾値未満も含めて全部**残し、passed（閾値以上）と
// adopted（上位 max_positions 件＝採用候補）の印を付ける。
//
// 全部残すのは事後検証のため——採用した銘柄だけを残すと、
// 「選ばなかった銘柄はどうなったか」と比べられなくなる（pkg/wbjp/evaluate）。
package history

import (
	"sort"

	"github.com/lovemoneyhotspring/jstock-go/pkg/wbcore/domain"
	corehistory "github.com/lovemoneyhotspring/jstock-go/pkg/wbcore/history"
	"github.com/lovemoneyhotspring/jstock-go/pkg/wbcore/settings"
	"github.com/shopspring/decimal"
)

// Kind は screen の履歴の種類名（ディレクトリ名）。
const Kind = "screen"

// MetaKeys は戦略が meta に残す数値のうち履歴に持つもの（無ければ null）。
var MetaKeys = []string{
	"dryup_ratio",
	"drawdown",
	"atr_ratio",
	"dollar_volume",
	"dryup",
	"rs",
	"trend",
	"liquid",
}

// ScreenColumns は screen 履歴の列。
//
// wbcore/history は Parquet に書ける型しか持たないので、Python 版の Int32 は
// Int64 に寄せている（DuckDB で読むときに型が揺れないようにするため）。
var ScreenColumns = screenColumns()

func screenColumns() []corehistory.Column {
	columns := []corehistory.Column{
		{Name: "rank", Type: corehistory.TypeInt64},
		{Name: "symbol", Type: corehistory.TypeString},
		{Name: "score", Type: corehistory.TypeFloat64},
		{Name: "passed", Type: corehistory.TypeBool},
		{Name: "adopted", Type: corehistory.TypeBool},
		{Name: "close", Type: corehistory.TypeFloat64},
		{Name: "reason", Type: corehistory.TypeString},
	}
	for _, key := range MetaKeys {
		columns = append(columns, corehistory.Column{Name: key, Type: corehistory.TypeFloat64})
	}
	return append(columns,
		corehistory.Column{Name: "entry_threshold", Type: corehistory.TypeFloat64},
		corehistory.Column{Name: "max_positions", Type: corehistory.TypeInt64},
		corehistory.Column{Name: "combiner", Type: corehistory.TypeString},
	)
}

// StoreFor は設定から wbjp の履歴ストアを作る。
func StoreFor(s *settings.AppSettings) *corehistory.Store {
	return corehistory.NewStore(s.WbjpHistoryDir())
}

// ScreenOptions は ScreenFrame の付随情報。
type ScreenOptions struct {
	// Threshold は買い建て候補とみなす合成スコアの下限。
	Threshold float64
	// MaxPositions は 1 回に採用する銘柄数の上限。
	MaxPositions int
	// Combiner は使った合成方式の名前。
	Combiner string
	// Meta は銘柄ごとの戦略の内訳（MetaKeys の値を拾う）。
	Meta map[string]map[string]any
	// Reasons は銘柄ごとの表示用の理由（無ければ CombinedSignal の Reason）。
	Reasons map[string]string
	// Close は銘柄の終値を引く。Meta に close があればそちらを優先する。
	Close func(symbol string) decimal.Decimal
}

// ScreenFrame は合成後の全銘柄をスコア順に並べ、閾値と採用枠の印を付ける。
//
// 並びは「スコアの高い順、同点なら銘柄コード順」。同点の順位が実行ごとに
// 入れ替わると、後から履歴を突き合わせたときに差分が読めなくなるため。
func ScreenFrame(combined []domain.CombinedSignal, opts ScreenOptions) corehistory.Frame {
	ordered := make([]domain.CombinedSignal, len(combined))
	copy(ordered, combined)
	sort.SliceStable(ordered, func(i, j int) bool {
		if ordered[i].Direction != ordered[j].Direction {
			return ordered[i].Direction > ordered[j].Direction
		}
		return ordered[i].Symbol < ordered[j].Symbol
	})

	rows := make([]map[string]any, 0, len(ordered))
	passedSoFar := 0
	for i, item := range ordered {
		meta := opts.Meta[item.Symbol]
		passed := item.Direction >= opts.Threshold
		if passed {
			passedSoFar++
		}

		reason := item.Reason
		if text, ok := opts.Reasons[item.Symbol]; ok && text != "" {
			reason = text
		}

		row := map[string]any{
			"rank":            int64(i + 1),
			"symbol":          item.Symbol,
			"score":           item.Direction,
			"passed":          passed,
			"adopted":         passed && passedSoFar <= opts.MaxPositions,
			"close":           closeOf(meta, opts.Close, item.Symbol),
			"reason":          reason,
			"entry_threshold": opts.Threshold,
			"max_positions":   int64(opts.MaxPositions),
			"combiner":        opts.Combiner,
		}
		for _, key := range MetaKeys {
			row[key] = corehistory.ToFloat(meta[key])
		}
		rows = append(rows, row)
	}
	return corehistory.NewFrame(ScreenColumns, rows)
}

// closeOf は meta の close を優先し、無ければ引き当て関数に聞く。
func closeOf(meta map[string]any, lookup func(string) decimal.Decimal, symbol string) any {
	if value := corehistory.ToFloat(meta["close"]); value != nil {
		return value
	}
	if lookup == nil {
		return nil
	}
	return corehistory.ToFloat(lookup(symbol))
}

File: src/wbjp/history.py
"""スイング売買のスクリーニング履歴（``wbjp screen`` の結果を積む）。

置き場は ``state/wbjp/history/screen/``（:class:`wbcore.history.HistoryStore`）。
1 回の ``screen`` が 1 ファイル。合成後の意見がある銘柄は**閾値未満も含めて全部**残し、
``passed``（閾値以上）と ``adopted``（上位 ``max_positions`` 件＝採用候補）の印を付ける。
"""

from __future__ import annotations

from collections.abc import Callable
from decimal import Decimal
from typing import TYPE_CHECKING, Any

import polars as pl

from wbcore.history import HistoryStore

if TYPE_CHECKING:
    from wbcore.domain.models import CombinedSignal
    from wbcore.settings import AppSettings

#: 戦略が ``meta`` に残す数値のうち履歴に持つもの（無ければ null）
META_KEYS = (
    "dryup_ratio",
    "drawdown",
    "atr_ratio",
    "dollar_volume",
    "dryup",
    "rs",
    "trend",
    "liquid",
)

SCREEN_SCHEMA: dict[str, Any] = {
    "rank": pl.Int32,
    "symbol": pl.Utf8,
    "score": pl.Float64,
    "passed": pl.Boolean,
    "adopted": pl.Boolean,
    "close": pl.Float64,
    "reason": pl.Utf8,
    **{key: pl.Float64 for key in META_KEYS},
    "entry_threshold": pl.Float64,
    "max_positions": pl.Int32,
    "combiner": pl.Utf8,
}


def store_for(settings: AppSettings) -> HistoryStore:
    return HistoryStore(settings.wbjp_history_dir)


def _num(value: object) -> float | None:
    """数値・Decimal を float に。None や型の違うものは null。"""
    if isinstance(value, int | float | Decimal):
        return float(value)
    return None


def screen_frame(
    combined: dict[str, CombinedSignal],
    meta: dict[str, dict[str, object]],
    reasons: dict[str, str],
    *,
    threshold: float,
    max_positions: int,
    combiner: str,
    close: Callable[[str], Decimal],
) -> pl.DataFrame:
    """合成後の全銘柄をスコア順に並べ、閾値と採用枠の印を付ける。"""
    ordered = sorted(combined.values(), key=lambda c: (-c.direction, c.symbol))
    rows: dict[str, list[Any]] = {key: [] for key in SCREEN_SCHEMA}
    passed_so_far = 0
    for rank, item in enumerate(ordered, start=1):
        m = meta.get(item.symbol, {})
        passed = item.direction >= threshold
        if passed:
            passed_so_far += 1
        rows["rank"].append(rank)
        rows["symbol"].append(item.symbol)
        rows["score"].append(float(item.direction))
        rows["passed"].append(passed)
        rows["adopted"].append(passed and passed_so_far <= max_positions)
        rows["close"].append(_num(m.get("close")) or _num(close(item.symbol)))
        rows["reason"].append(reasons.get(item.symbol, item.reason))
        for key in META_KEYS:
            rows[key].append(_num(m.get(key)))
        rows["entry_threshold"].append(float(threshold))
        rows["max_positions"].append(max_positions)
        rows["combiner"].append(combiner)
    return pl.DataFrame(rows, schema=SCREEN_SCHEMA)


__all__ = ["META_KEYS", "SCREEN_SCHEMA", "screen_frame", "store_for"]
